wasm-chain-integration/src/v0/ffi.rs

```diff
@@ -188,8 +182,10 @@ unsafe extern "C" fn validate_and_process_v0(
 /// This function is safe provided the supplied pointer is not null and is
 /// constructed with [Arc::into_raw].
 unsafe extern "C" fn artifact_v0_free(artifact_ptr: *const ArtifactV0) {
-    // decrease the reference count
-    Arc::from_raw(artifact_ptr);
+    if !artifact_ptr.is_null() {
```

Comment:
Nope. `Arc::from_raw(Box::into_raw(Box::new(Artifact...)))` will explode.

This is only safe if the pointer is constructed with Arc::into_raw, and each ::into_raw has at most one ::from_raw (otherwise double-free).